### PR TITLE
918/url encode utils

### PR DESCRIPTION
**Pull request by alfetopito**
Added encode/decode helper functions to address https://github.com/gnosis/dex-react/issues/918

Published RC to npm `v0.2.3-RC.0`

### CODE REVIEW

File: test/utils/urlEncoding/encodeSymbol.spec.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { encodeSymbol } from '../../../src/utils'
+
+describe('Encoding symbol', () => {
+  test('A-Z symbol', () => {
+    expect(encodeSymbol('DAI')).toEqual('DAI')
+  })
+
+  test('Special chars symbol', () => {
+    expect(encodeSymbol('Token /21 - 55? ** |-=+#@%^&`\'"')).toEqual(
```
Comment (Velenir) [inline test/utils/urlEncoding/encodeSymbol.spec.ts]: Why `.toEqual` some times and `.toBe` other times?

File: src/utils/urlEncoding.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * Decodes symbol which is URL encoded.
+ * Trims trailing and leading spaces.
+ *
+ * @param symbol URL encoded symbol
+ */
+export function decodeSymbol(symbol: string): string {
+  return decodeURIComponent(symbol).trim()
+}
+
+const dashRegex = new RegExp(/-/, 'g')
```
Comment (Velenir) [inline src/utils/urlEncoding.ts]: This is a minor annoyance, maybe even personal minor annoyance, but when source of RegExp isn't dynamic, don't create it through a constructor. It copies unnecessarily
Also see gnosis/dex-react#933
```suggestion
const dashRegex = /-/g
```

**Review by Velenir — APPROVED** [review]

File: src/utils/urlEncoding.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+/**
+ * Syntactic sugar for encoding Token objects where `symbol` is an optional property.
+ * When there's no symbol, uses token address instead.
+ *
+ * @param token Object containing address and optionally symbol
+ */
+export function encodeTokenSymbol(token: { symbol?: string; address: string }): string {
+  const tokenSymbol = token.symbol || token.address
```
Comment (Velenir) [inline src/utils/urlEncoding.ts]: no need to encode address, not that it's wrong

File: test/utils/urlEncoding/encodeSymbol.spec.ts (hunk 1 above)
Comment (alfetopito) [reply test/utils/urlEncoding/encodeSymbol.spec.ts]: Because I can't make up my mind :D

File: src/utils/urlEncoding.ts (hunk 3 above)
Comment (alfetopito) [reply src/utils/urlEncoding.ts:31]: 100% support you on that

Comment (alfetopito) [reply src/utils/urlEncoding.ts]: ok

File: test/utils/urlEncoding/decodeSymbol.spec.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { decodeSymbol } from '../../../src/utils'
+describe('Decoding symbol', () => {
+  test('A-Z symbol', () => {
+    expect(decodeSymbol('TNK')).toBe('TNK')
+  })
+
+  test('Special chars', () => {
+    expect(decodeSymbol("Token%20%2F21%20%_%2055%3F%20**%20%7C%_%3D%2B%23%40%25%5E%26%60'%22")).toBe(
+      'Token /21 - 55? ** |-=+#@%^&`\'"',
```
Comment (anxolin) [inline test/utils/urlEncoding/decodeSymbol.spec.ts:9]: Amazing token name!
This is defensive programing 🤘

File: test/utils/urlEncoding/decodeSymbol.spec.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    )
+  })
+  test('Dash in the symbol', () => {
+    expect(decodeSymbol('TN%_HUI')).toBe('TN-HUI')
```
Comment (anxolin) [inline test/utils/urlEncoding/decodeSymbol.spec.ts]: Yeah baby! not pretty, but what the hell, I would be surprised if we ever see a real token with a dash. 
I think no-one would care or notice that we encode the the dash in a non standard way, so I'm all for it!

File: test/utils/urlEncoding/decodeSymbol.spec.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  })
+  test('Leading and trailing spaces', () => {
+    expect(decodeSymbol('%20SPACES%20HERE%20%20')).toBe('SPACES HERE')
+  })
+  test('Unicode chars', () => {
+    expect(decodeSymbol('Oh%20%F0%9F%92%BB%20%F0%9F%A7%B5%20%F0%9F%94%A5')).toBe('Oh 💻 🧵 🔥')
```
Comment (anxolin) [inline test/utils/urlEncoding/decodeSymbol.spec.ts:19]: This is more realistic, actually, I'm going to mint a token myself with emojis

**Review by anxolin — APPROVED** [review]

File: test/utils/urlEncoding/decodeSymbol.spec.ts (hunk 6 above)
Comment (Velenir) [reply test/utils/urlEncoding/decodeSymbol.spec.ts:19]: Just don't call it Emo Token :joy:

Comment (alfetopito) [reply test/utils/urlEncoding/decodeSymbol.spec.ts:19]: I dare you!

File: test/utils/urlEncoding/decodeSymbol.spec.ts (hunk 5 above)
Comment (alfetopito) [reply test/utils/urlEncoding/decodeSymbol.spec.ts]: As we've seen, browsers do not like encoding that's not really an encoding:

screenshot_2020-04-17_09-24-39

File: src/utils/urlEncoding.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
+// We cannot have a dash in the symbol because:
+// 1. We use a `-` to separate token symbols in the URL
```
Comment (anxolin) [inline src/utils/urlEncoding.ts]: unimportant, but, maybe we could say, we use it to name the market, for example "A-B", where A is the sell token and B is the buy token

File: src/utils/urlEncoding.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * Encodes given symbol into a URL friendly string.
+ * Additionally, encodes chars that might be ambiguous in the URL
+ * such as a dash `-`
+ *
+ * @param symbol Symbol to URL encode
+ */
+export function encodeSymbol(symbol: string): string {
+  const sanitizedSymbol = symbol
+    .trim() // Your token has leading/trailing spaces? Bad luck
+    .replace(dashRegex, encodedDashSymbol) // Your token has dashes? you won't see them for awhile
```
Comment (anxolin) [inline src/utils/urlEncoding.ts:24]: 😂nice comments

File: src/utils/urlEncoding.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// We cannot have a dash in the symbol because:
+// 1. We use a `-` to separate token symbols in the URL
+// 2. Tokens can have dashes in their symbols
+// 3. Even if we encode dashes to `%2D`, browsers auto convert them back to `-`
+// For these reasons, we are picking a symbol that is not ambiguous with a dash
+// and it have very little chance to be used as part of a token symbol.
+// Meet, the question circle dash ⊝ https://www.fileformat.info/info/unicode/char/229d/index.htm
```
Comment (Velenir) [inline src/utils/urlEncoding.ts]: Brb, creating a Token `T⊝KEN`

**Review by Velenir — APPROVED** [review]

**Review by anxolin — APPROVED** [review]

File: src/utils/urlEncoding.ts (hunk 9 above)
Comment (alfetopito) [reply src/utils/urlEncoding.ts]: You are right, better not advertise that because of people like you!

:D

File: src/utils/urlEncoding.ts (hunk 8 above)
Comment (alfetopito) [reply src/utils/urlEncoding.ts:24]: ;)

File: src/utils/urlEncoding.ts (hunk 7 above)
Comment (alfetopito) [reply src/utils/urlEncoding.ts]: Updated

**Pull request merged**